File: BlackJack.cpp
//John Kutbay
//BlackJack Lab
//COSC102
//5/27/22
//In this lab, I will create a program that simulates blackjack for the user.

#include <iostream>
#include <string>
#include <iomanip>
#include <cstdlib>
#include <ctime>

using namespace std;

//This is the declaration of my functions and the constant size of a deck.

const int numCard = 52;

string DealCard(int cards[numCard], int player);

int GetRandom(int min, int max);

void InitializeCards(int cards[numCard]);

int ScoreHand(int cards[numCard], int player);

int main() {

//I started by initializing my my cards and initializing variables.
	int cards[numCard];
	InitializeCards(cards);
	srand(time(NULL));
	string hitStand = "";
	const string cardValue[13] = {"A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K"};
	const string cardSuit[4] = {"C", "D", "H", "S"};
	string dealersCards = "";
	string usersCards = "";

//This section is where I dealt the initial cards and scored the results.
	cout << "Dealer has cards: ";
	cout << setw(20) << left;
	DealCard(cards,1);
	DealCard(cards,1);
	for (int i = 0;i < numCard;i++) {
		if (cards[i] == 1) {
			dealersCards += cardValue[i % 13] + cardSuit[i / 13] + " ";
		}
	}
	cout << dealersCards;
	cout << "(" << ScoreHand(cards, 1) << ")" << '\n';

	cout << "You have cards  : ";
	cout << setw(20) << left;
	DealCard(cards, 2);
	DealCard(cards, 2);
	for (int i = 0;i < numCard;i++) {
		if (cards[i] == 2) {
			usersCards += cardValue[i % 13] + cardSuit[i / 13] + " ";
		}
	}
	cout << usersCards;
	cout << "(" << ScoreHand(cards, 2) << ")" << '\n';

	cout << "Hit or stand    ? ";
	cin >> hitStand;
	cout << '\n';
	
//This section is where I ran a series of if statements and loops to get input
//if the user wants to hit or stand.
	while (true) {
		if (hitStand == "hit") {
			cout << "Dealer has cards: ";
			cout << setw(20) << left;
			cout << dealersCards;
			cout << "(" << ScoreHand(cards, 1) << ")";
			cout << '\n';

			cout << "You have cards  : ";
			cout << setw(20) << left;
			usersCards += DealCard(cards, 2) + " ";
			cout << usersCards;
			cout << "(" << ScoreHand(cards, 2) << ")";
			cout << '\n';

			if (ScoreHand(cards, 2) > 21) {
				cout << "Player busts, dealer wins!\n";
				return -1;
			} else if (ScoreHand(cards, 2) == 21) {
				cout << "Player wins!\n";
				return -1;
			} else {
				cout << "Hit or stand    ? ";
				cin >> hitStand;
				cout << '\n';
			}		
		}else if (hitStand == "stand") {

//In this section, I used the restriction of scores for the dealer
//to determine if the dealer should hit or stand.

			if (ScoreHand(cards, 1) < 18) {
				cout << "Dealer hits     : ";
				cout << setw(20) << left;
				dealersCards += DealCard(cards, 1) + " ";
				cout << dealersCards;
				cout << "(" << ScoreHand(cards, 1) << ")";
				cout << '\n';

			} else if (ScoreHand(cards, 1) == 21) {
				cout << "Dealer wins!\n";
				return -1;
			} else if (ScoreHand(cards, 1) > 21) {
				cout << "Dealer busts, player wins!\n";
				return -1;
			} else if (ScoreHand(cards, 1) >= 18 || ScoreHand(cards, 1) < 21) {
				cout << "Dealer Stands  : ";
				cout << setw(20) << left;
				cout << dealersCards;
				cout << "(" << ScoreHand(cards, 1) << ")";
				cout << '\n' << '\n';
				break;
			}
		}
	}

//This is where I scored the results if neither dealer or player hit 21 or busted.
	if (ScoreHand(cards, 1) > ScoreHand(cards, 2)) {
		cout << "Dealer wins!\n";
	} else if (ScoreHand(cards, 2) > ScoreHand(cards, 1)) {
		cout << "Player wins!\n";
	} else if (ScoreHand(cards, 1) == ScoreHand(cards, 2)) {
		cout << "Player and dealer draw.\n";
	}
}

//This function deals a card to whoever is called.

string DealCard(int cards[numCard], int player) {

	int randomNumber;
	string card = "";
	const string cardValue[13] = {"A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K"};
	const string cardSuit[4] = {"C", "D", "H", "S"};
	randomNumber = GetRandom(0, 51);
	while (cards[randomNumber] != 0) {
		randomNumber = GetRandom(0, 51);
	}
	cards[randomNumber] = player;
	card = cardValue[randomNumber % 13] + cardSuit[randomNumber / 13];
	return card;
}	

//This function returns a random number between the stated values.

int GetRandom(int min, int max) {
	
	return min + rand() % (max - min + 1);
}

//This function initializes an array with zeros.

void InitializeCards(int cards[numCard]) {
	for (int i = 0;i < numCard;i++) {
		cards[i] = 0;
	}
}

//This function scores a player's scores.

int ScoreHand(int cards[numCard], int player) {
	
	int scoreValue = 0;

	for (int i = 0; i < numCard; i++) {
		if (cards[i] == player) {
			if ((i % 13) < 10) {
				scoreValue += ((i % 13) + 1);
			} else if ((i % 13) >= 10) {
				scoreValue += 10;
			}
		}
	}
	return scoreValue;
}

File: BlackJack.py
# John Kutbay

import random

test = "Hello"

def init():
    global cards
    cards = []
    global users
    global dealers
    users = []
    dealers = []
    global cardV
    global cardS
    cardV = ["A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K"]
    cardS = ["P", "S", "C", "H"]
    for i in range(0,52):
        cards.append(0)

def deal(num, player):
  for i in range(0,num):
    cardN = random.randint(0,9)
    while cards[cardN] != 0:
      cardN = random.randint(0,9)
    cards[cardN] = player
    if player == 1:
      users.append(cardV[int(cardN % 13)] + cardS[int(cardN / 13)])
    else:
      dealers.append(cardV[int(cardN % 13)] + cardS[int(cardN / 13)])
    
def score(player):
  scoreVal = 0

  for i in range(52):
    if cards[i] == player:
      if (i%13) < 10 and (i%13) != 0:
        scoreVal += (i%13)+1
      elif (i%13) >= 10:
        scoreVal += 10
      else:
        if (scoreVal + 11 > 21):
          scoreVal += 1
        else:
          scoreVal += 11
  return scoreVal




init()

print("WELCOME TO BLACKJACK!")
print('\n')

deal(2,1)
print("You have cards: ", end = '')
for i in range(len(users)):
  print(users[i] + " ", end = '')
print("(" + str(score(1)) + ")", end = '\n')

if score(1) == 21:
  print("Player wins!\n")
  exit(0)

deal(2, 2)
print("Dealer has cards: ", end = '')
for i in range(len(dealers)):
  print(dealers[i] + " ", end = '')
print("(" + str(score(2)) + ")", end = '\n')

if score(2) == 21:
  print("Dealer wins!\n")
  exit(0)

while True:

  hitStand = input("Hit or Stand?(h or s) ")
  
  if hitStand == 'h':
    deal(1,1)
    print("You have cards: ", end = '')
    for i in range(len(users)):
      print(users[i] + " ", end = '')
    print("(" + str(score(1)) + ")", end = '\n')

    if score(1) > 21:
      print("Player busts, Dealer wins!\n")
      exit(0)
    elif score(1) == 21:
      print("Player wins!\n")
      exit(0)

  elif hitStand == 's':

    while score(2) < 17 or score(2) < score(1):
      print("Dealer Hits! ", end = '')
      deal(1,2)
      for i in range(len(dealers)):
        print(dealers[i] + " ", end = '')
      print("(" + str(score(2)) + ")", end = '\n')

    if score(2) == 21:
      print("Dealer wins!\n")
      exit(0)
    elif score(2) > 21:
      print("Dealer busts!\n")
      exit(0)
    else:
      print("Dealer Stands.\n")
      break

if score(1) == score(2):
  print("Game is a draw!\n")
  exit(0)









